File: app/src/main/java/com/icostel/arhitecturesample/repository/UserLogInHandler.java
package com.icostel.arhitecturesample.repository;

import com.icostel.arhitecturesample.api.UserApiService;
import com.icostel.arhitecturesample.api.model.SignInResponse;
import com.icostel.arhitecturesample.api.utils.ApiErrorResponse;
import com.icostel.arhitecturesample.api.utils.ApiResponse;
import com.icostel.arhitecturesample.api.utils.ApiSuccessResponse;
import com.icostel.arhitecturesample.api.utils.SessionStore;
import com.icostel.arhitecturesample.api.utils.SignInStatus;
import com.icostel.arhitecturesample.utils.AppExecutors;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

import androidx.annotation.NonNull;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

@Singleton
public class UserLogInHandler {

    private UserApiService userApiService;
    private AppExecutors appExecutors;
    private SessionStore sessionStore;

    @Inject
    UserLogInHandler(AppExecutors executors, UserApiService userApiService, SessionStore sessionStore) {
        this.userApiService = userApiService;
        this.appExecutors = executors;
        this.sessionStore = sessionStore;
    }

    // other API calls will use the token obtained from user sign in and stored in session store
    public void signInUser(@NonNull String userEmail, @NonNull String userPass,
                           @NonNull OnUserSignInResultListener listener) {
        // add a delay to simulate a bad network or something, a little nasty to read tho
        Schedulers.io().createWorker().schedule(() ->
                appExecutors.networkIO().execute(() -> {
                    try {
                        manageResponse(ApiResponse.create(userApiService.signInUser(userEmail, userPass).execute()), listener);
                    } catch (Exception e) {
                        Timber.e(e);
                        listener.onUserSignInResult(SignInStatus.Error());
                    }
                }), 3000, TimeUnit.MILLISECONDS);
    }

    private void manageResponse(ApiResponse<SignInResponse> response, @NonNull OnUserSignInResultListener listener) {
        if (response instanceof ApiSuccessResponse) {
            SignInResponse signInResponse = ((ApiSuccessResponse<SignInResponse>) response).getBody();
            if (signInResponse.getSuccess()) {
                sessionStore.setUserSessionToken(signInResponse.getToken());
                listener.onUserSignInResult(SignInStatus.Success());
                Timber.d("===Sign in success: " + ((ApiSuccessResponse) response).getBody());
            } else {
                listener.onUserSignInResult(SignInStatus.Error());
                Timber.d("===Sign in error: " + ((ApiErrorResponse) response).getErrorMessage());
            }
        } else if (response instanceof ApiErrorResponse) {
            listener.onUserSignInResult(SignInStatus.Error());
            Timber.d("===Sign in error: " + ((ApiErrorResponse) response).getErrorMessage());
        }
    }

    public interface OnUserSignInResultListener {
        void onUserSignInResult(SignInStatus status);
    }
}

File: app/src/main/java/com/icostel/arhitecturesample/repository/UserLogInHandler.kt
package com.icostel.arhitecturesample.repository

import android.os.Handler
import com.icostel.arhitecturesample.api.Status
import com.icostel.arhitecturesample.api.UserApiService
import com.icostel.arhitecturesample.api.model.SignInResponse
import com.icostel.arhitecturesample.api.session.SessionStore
import com.icostel.arhitecturesample.api.utils.ApiErrorResponse
import com.icostel.arhitecturesample.api.utils.ApiResponse
import com.icostel.arhitecturesample.api.utils.ApiSuccessResponse
import com.icostel.commons.utils.AppExecutors
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class UserLogInHandler @Inject
internal constructor(private val appExecutors: AppExecutors,
                     private val userApiService: UserApiService,
                     private val sessionStore: SessionStore) {

    // other API calls will use the token obtained from user sign in and stored in session store
    fun signInUser(userEmail: String, userPass: String,
                   listener: OnUserSignInResultListener) {
        Handler().postDelayed({
            appExecutors.networkIO().execute {
                try {
                    manageResponse(ApiResponse.create(userApiService.signInUser(userEmail, userPass).execute()), listener)
                } catch (e: Exception) {
                    Timber.e(e)
                    listener.onUserSignInResult(Status.error())
                }
            }
        }, 2000L)
    }

    private fun manageResponse(response: ApiResponse<SignInResponse>, listener: OnUserSignInResultListener) {
        when (response) {
            is ApiSuccessResponse<*> -> {
                val (success, _, token) = (response as ApiSuccessResponse<SignInResponse>).body
                if (success) {
                    sessionStore.setUserToken(token)
                    listener.onUserSignInResult(Status.success())
                    Timber.d("===Sign in success: " + (response as ApiSuccessResponse<*>).body)
                } else {
                    listener.onUserSignInResult(Status.error())
                    sessionStore.setUserToken(null)
                    Timber.d("===Sign in error: " + (response as ApiErrorResponse<*>).errorMessage)
                }
            }
            is ApiErrorResponse<*> -> {
                listener.onUserSignInResult(Status.error())
                Timber.d("===Sign in error: " + (response as ApiErrorResponse<*>).errorMessage)
            }
        }
    }

    interface OnUserSignInResultListener {
        fun onUserSignInResult(status: Status)
    }
}
